src/server.c

```diff
@@ -572,27 +572,33 @@ static void monitor_cb(uv_prepare_t *monitor)
 {
 	struct dqlite_node *d = monitor->data;
 	int state = raft_state(&d->raft);
-	/*
 	queue *head;
 	struct conn *conn;
-	*/
+	struct gateway *g;
 
 	if (state == RAFT_UNAVAILABLE) {
 		return;
 	}
 
-	/* TODO: we should shutdown clients that are performing SQL requests,
-	 * but not the ones which are doing management-requests, such as
-	 * transfer leadership.  */
-	/*
 	if (d->raft_state == RAFT_LEADER && state != RAFT_LEADER) {
+		tracef("leadership lost, close leader connections");
 		QUEUE__FOREACH(head, &d->conns)
 		{
 			conn = QUEUE__DATA(head, struct conn, queue);
-			conn__stop(conn);
+			g = &conn->gateway;
+			if (g == NULL || g->leader == NULL) {
+				continue;
```

Comment:
Why are you checking for ```g->leader``` also? In case ```g->leader``` is ```NULL``` this should be a connection that is handling management requests no? So in that case we treat the connection differently and not close it? Or perhaps I'm confused.

Reply:
The main point of this PR was to close the sqlite leader connection so that sqlite cleans up its state from an ongoing sql request resulting from a leadership loss. I reasoned that connections without a `g->leader` could stay open as they don't have an open sqlite leader connection and are safe to be reused by the sql driver.

Reply:
I see.

I believe that there are no cases where we mix using a connection both for SQL and management requests. So perhaps for sake of simplicity you could just check that and simplify the code a bit.

Reply:
Hmm yes, you're right, the go client doesn't mix them. They are mixed in the C unit tests though, will adapt and just close any connection that has a `gateway->leader`, then we don't need that delayed close logic at all.